### PR TITLE
Implement a service for operating location entity

### LINKED ISSUE #463: Implement a service for operating Location entity
~The task is blocked by:~
~- [x] #454~

### Overview
Implement a service to operate the `Location` data model.

### Action Items

- [ ] Add `location.service.js` into `backend/services` folder. 
- [ ] Implement `add(locations)` function without returning value. Features:
* `locations` parameter is an array of non-empty strings with at least one element. This should be validated.
* the function adds elements to the `Locations` DB collection skipping duplicates.
- [ ] Implement `getAll()` function that returns all `Location` values as an array of strings.
- [ ] For both functions wrap DB call using `try/catch` with throwing [DatabaseError](https://github.com/hackforla/VRMS/blob/development/backend/errors/database.error.js) in case of any error.
- [ ] Cover implemented functionality with tests.

### Resources/Instructions
[Service example](https://github.com/hackforla/VRMS/blob/development/backend/services/modificationLog.service.js)
[Service test example](https://github.com/hackforla/VRMS/blob/development/backend/services/modificationLog.service.test.js)

### PR DESCRIPTION
**Pull request by akibrhast**
closes #463

This PR implements the below functionality.The host of the functionality is implemented using the intricacies of mongoose `findOneAndUpdate`

- [x] Add `location.service.js` into `backend/services` folder. 
- [x] Implement `add(locations)` function without returning value. Features:
* `locations` parameter is an array of non-empty strings with at least one element. This should be validated.
* the function adds elements to the `Locations` DB collection skipping duplicates.
- [x] Implement `getAll()` function that returns all `Location` values as an array of strings.
- [x] For both functions wrap DB call using `try/catch` with throwing [DatabaseError](https://github.com/hackforla/VRMS/blob/development/backend/errors/database.error.js) in case of any error.
- [x] Cover implemented functionality with tests.

It implements the following tests
 - Add empty array should throw database error
 - Add an array of location strings sucessfully adds to database
 - Add duplicate string to location array should skip silently
 - Add array with whitespace string should throw database error
 - Should return an empty array if no location data found
 - Should return an array of location string if data exists

### CODE REVIEW

**Comment by peteplass** [comment]
Talked about in meeting, Alex to review

File: backend/services/globalConfig.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"mongoUri":"mongodb://127.0.0.1:49438/cd3e8885-3b20-4ea8-8d86-0c2009ce8001?"}
```
Comment (drubgrubby) [inline backend/services/globalConfig.json]: Please remove from pull request.

**Review by drubgrubby — DISMISSED** [review]
@akibrhast -> This looks

File: backend/models/dictionaries/location.model.js [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
+        type: [{
+            type: String,
+            required: true,
+            validate: {
+                validator: function(v) {
+                  return v.trim().length > 0;
+                },
+                message: props => `An empty string is not allowed`
+            }
+        }],
```
Comment (alex-anakin) [inline backend/models/dictionaries/location.model.js]: please, move validation into service

File: backend/services/location.service.test.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    
+    const isArrayUnique = arr => Array.isArray(arr) && new Set(arr).size === arr.length;
+    await locationService.add(['loc1', 'loc2', 'loc3']);
+    const locationsArray = ['loc3', 'loc4', 'loc5'];
+    await locationService.add(locationsArray);
+    const savedLocations = await locationService.getAll();
+
+    expect(isArrayUnique(savedLocations)).toBeTruthy();
+    expect(savedLocations.length).toEqual(5);
+    expect(savedLocations).toEqual(expect.arrayContaining(locationsArray));
```
Comment (alex-anakin) [inline backend/services/location.service.test.js]: please check out that result contains locations 1-5

**Review by alex-anakin — CHANGES_REQUESTED** [review]

**Comment by akibrhast** [comment]
### Progress
None since my last comment

### Blockers
busy with hfla/website

### Availability
pretty free this week

### ETA
by friday

File: backend/models/dictionaries/location.model.js (hunk 2 above)
Comment (akibrhast) [reply backend/models/dictionaries/location.model.js]: validation moved to service

File: backend/services/location.service.test.js (hunk 3 above)
Comment (akibrhast) [reply backend/services/location.service.test.js]: result contains 1-5

File: backend/globalConfig.json [hunk 4]
```diff
@@ -1 +1 @@
-{"mongoUri":"mongodb://127.0.0.1:3170/jest?","mongoDBName":"jest"}
\ No newline at end of file
+{"mongoUri":"mongodb://127.0.0.1:3170/jest?","mongoDBName":"jest"}
```
Comment (akibrhast) [inline backend/globalConfig.json]: This file was restored to original.

File: backend/services/globalConfig.json (hunk 1 above)
Comment (akibrhast) [reply backend/services/globalConfig.json]: removed

File: backend/services/location.service.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/* eslint-disable func-names */
+const { Location } = require('../models/dictionaries/location.model');
+const DatabaseError = require('../errors/database.error');
+
+const LocationService = {};
+
+function validateLocations(locationsarray){
+
+  if(!Array.isArray(locationsarray)){
+    throw "locations array must be of type array"
```
Comment (alex-anakin) [inline backend/services/location.service.js]: please improve the message to "locations input must be of type array"

File: backend/services/location.service.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  }
+  if(locationsarray.length <= 0){
```
Comment (alex-anakin) [inline backend/services/location.service.js]: please fix the condition. It should be `locationsarray.length === 0`

File: backend/services/location.service.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    throw "locations array must have at least one element";
+  }
+  if( !locationsarray.every(item => (typeof item === "string")) ){
+    throw "locations array must be an array of strings"
+  }
+  if( !locationsarray.every( (element) => { if(element){return true} } )){
+    throw "locations parameter is an array of non-empty strings"
```
Comment (alex-anakin) [inline backend/services/location.service.js]: please improve the message to "locations array should not contain empty strings"

File: backend/services/location.service.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/* eslint-disable func-names */
+const { Location } = require('../models/dictionaries/location.model');
+const DatabaseError = require('../errors/database.error');
+
+const LocationService = {};
+
+function validateLocations(locationsarray){
```
Comment (alex-anakin) [inline backend/services/location.service.js:8]: in all checks please throw [ValidationError](https://github.com/hackforla/VRMS/blob/development/backend/errors/validation.error.js) that contains specific message

**Review by alex-anakin — CHANGES_REQUESTED** [review]
Please fix comments

File: backend/services/location.service.js (hunk 8 above)
Comment (akibrhast) [reply backend/services/location.service.js:8]: all checks now throwing validation error

File: backend/services/location.service.js (hunk 7 above)
Comment (akibrhast) [reply backend/services/location.service.js]: messages improved as stated

File: backend/services/location.service.js (hunk 6 above)
Comment (akibrhast) [reply backend/services/location.service.js]: condition fixed

File: backend/services/location.service.js (hunk 5 above)
Comment (akibrhast) [reply backend/services/location.service.js]: message improved

**Review by alex-anakin — APPROVED** [review]

**Pull request merged**